File: oi/oi1.hs
import Control.Monad (forM_, when)
import Data.Array.ST
    ( getElems, readArray, writeArray, MArray(newArray), STUArray )
import Data.Array.Unboxed
import Data.Maybe (isNothing)
import Control.Monad.ST (runST, ST)
import Data.List (foldl')
import Data.STRef
import Control.Monad.Cont
import Control.Arrow ((>>>))

type Edge = (Int, Int, Int)

main :: IO ()
main =
  interact $
    lines
      >>> map (words >>> map read)
      >>> process

process :: [[Int]] -> String
process ([t,n]:xs) = if good then unwords (map show res) ++ "\n" else "neexistuje\n"
    where (good, res) = solve t n (map (\[a, b, d] -> (a-1, b-1, -d)) (take n xs))



solve :: Int -> Int -> [Edge] -> (Bool, [Int])
solve t n xs =
    runST $ do
        dist <- newArray (0, t-1) 0 :: ST s (STUArray s Int Int)
        changed <- newSTRef True
        let update =
                forM xs $ \(a, b, d) -> do
                    da <- readArray dist a
                    db <- readArray dist b
                    when (da + d < db) $ writeArray dist b (da + d)
                    when (da + d < db) $ writeSTRef changed True

        forM_ [1..t] $ \_ ->
          do
            writeSTRef changed False
            update

        finalChanged <- readSTRef changed
        if finalChanged
        then return (False, [])
        else do
          res <- getElems dist
          return (True, res)

File: oi/oi1.py
def main():
    t, n = map(int, input().split())

    edges = [(0, 0, 0)]*n

    for i in range(n):
        a, b, d = map(int, input().split())
        a -= 1
        b -= 1
        edges[i] = (a, b, -d)


    changed = False
    dist = [0]*t

    for i in range(t):
        changed = False
        for e in edges:
            if dist[e[1]] - dist[e[0]] > e[2]:
                changed = True
                dist[e[1]] = dist[e[0]] + e[2]
        if not changed:
            break
    
    if changed:
        print("neexistuje")
    else:
        for i in range(t):
            print(dist[i], end = '\n' if i == t-1 else ' ')

main()

